### ISSUE TITLE
Default graceful shutdown is different between Armeria Server and Spring integration.

### ISSUE THREAD

**Issue by minwoox** [line/armeria#3283]
If a user starts a server manually(i.e. Server.start()),
the graceful shutdown is disabled.
https://github.com/line/armeria/blob/armeria-1.3.0/core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java#L138
However, if the user uses Spring auto integration, the graceful shutdown is enabled.
https://github.com/line/armeria/blob/master/spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaSettings.java#L287-L302
https://github.com/line/armeria/blob/master/spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/internal/spring/ArmeriaConfigurationUtil.java#L101

I don't remember why we have different settings for that.
Is there anyone who knows about this?

**Comment by heowc** [comment]
As you may know, it seems to have been added in [this PR](https://github.com/line/armeria/pull/494). However, it is not possible to know exactly because there is no content mentioned separately during this PR.

See https://github.com/line/armeria/pull/494

**Comment by anuraaga** [comment]
@minwoox I guess it's because we would generally expect spring autoconfigure to provide a best-practices out-of-box experience, while programatic configuration is generally expected to be more explicit.

**Comment by minwoox** [comment]
Yeah, that makes sense. It's the same context that Spring autoconfigure provides the health check service and metric collecting by default. Got it. Thanks!

**Comment by ikhoon** [comment]
>  the graceful shutdown is disabled.

I am not sure this is a sensible default. I prefer to give an appropriate graceful timeout.
When I was working on http4s-armeria, the graceful timeout for http4s is 30 seconds.
https://github.com/http4s/http4s/blob/main/server/src/main/scala/org/http4s/server/package.scala#L49-L50

**Comment by anuraaga** [comment]
@ikhoon Graceful shutdown can slow down non-production cases like unit tests in some cases. Maybe personal preference, but opting-in seems to make sense, especially given there's a lot of configuration needed for a production server.

https://armeria.dev/docs/advanced-production-checklist

I guess it would be nice to add graceful shutdown setting to that list though XD

**Comment by ikhoon** [comment]
> Graceful shutdown can slow down non-production cases like unit tests in some cases.

I am also worried about that point. I thought the graceful timeout of `ServerExtension` and `ServerRule` could be disabled for testing.

> armeria.dev/docs/advanced-production-checklist
> I guess it would be nice to add graceful shutdown setting to that list though XD

👍 Yeah. Graceful shutdown is an optional but essential feature in production mode.

**Comment by trustin** [comment]
I agree with @anuraaga :-)

**Comment by chris-ryan-square** [comment]
I was searching the documentation for details about server shutdown and the grace period etc.  There's nothing there that I can see (point me if I missed something!).  I think a small section in the docs about the shutdown process, grace period, listener state hooks (whenStopping, etc) and implications or limitations of it all would be a very useful addition.  Especially since I was not expecting the graceful shutdown to be disabled if the server is started manually.

**Comment by ikhoon** [comment]
>  I think a small section in the docs about the shutdown process, grace period, listener state hooks (whenStopping, etc) and implications or limitations of it all would be a very useful addition.

Nice suggestion! We should document for them.
